quickwit/quickwit-storage/src/opendal_storage/base.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::fmt;
+use std::ops::Range;
+use std::path::Path;
+
+use async_trait::async_trait;
+use bytesize::ByteSize;
+use opendal::Operator;
+use quickwit_common::uri::Uri;
+use tokio::io::{AsyncRead, AsyncWriteExt};
+
+use crate::storage::SendableAsync;
+use crate::{
+    BulkDeleteError, OwnedBytes, PutPayload, Storage, StorageError, StorageErrorKind,
+    StorageResolverError, StorageResult,
+};
+
+/// OpenDAL based storage implementation.
+/// # TODO
+///
+/// - Implement REQUEST_SEMAPHORE to control the concurrency.
+/// - Implement STORAGE_METRICS for metrics.
+/// - Add multipart_policy to control write at once or via multiple.
+#[derive(Clone)]
+pub struct OpendalStorage {
+    uri: Uri,
+    op: Operator,
+}
+
+impl fmt::Debug for OpendalStorage {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter
+            .debug_struct("OpendalStorage")
+            .field("operator", &self.op.info())
+            .finish()
+    }
+}
+
+impl OpendalStorage {
+    /// Create a new google cloud storage.
+    pub fn new_google_cloud_storage(
+        uri: Uri,
+        cfg: opendal::services::Gcs,
+    ) -> Result<Self, StorageResolverError> {
+        let op = Operator::new(cfg)?.finish();
+        Ok(Self { uri, op })
+    }
+}
+
+#[async_trait]
+impl Storage for OpendalStorage {
+    async fn check_connectivity(&self) -> anyhow::Result<()> {
+        self.op.check().await?;
+        Ok(())
+    }
+
+    /// # TODO
+    ///
+    /// We can implement something like `multipart_policy` determine whether to use copy.
```

Comment:
So opendal supports multipart upload? What does the API look like?

Reply:
> So opendal supports multipart upload? What does the API look like?

OpenDAL provides multipart support through the `Writer` interface. Each invocation of `writer.write()` triggers a multipart upload. Users can adjust internal behavior as follows:

- `op.writer_with(path).buffer(buffer_size)` set the buffer size of each part.
- `op.writer_with(path).concurrent(concurrent)` set the concurrent number to part uploading.

Reply:
awesome thank you!